nemo/collections/nlp/models/language_modeling/megatron/tokens_encoder_decoder.py

```diff
@@ -0,0 +1,282 @@
+    """
+
+    def __init__(self, mpu_vocab_size, parallel_output):
+        super(MegatronTokensHead, self).__init__()
+
+        self.bias = torch.nn.Parameter(torch.zeros(mpu_vocab_size))
+        self.bias.model_parallel = True
+        self.bias.partition_dim = 0
+        self.bias.stride = 1
+        self.parallel_output = parallel_output
+
+    def forward(self, hidden_states, word_embeddings_weight):
+        output = parallel_lm_logits(hidden_states, word_embeddings_weight, self.parallel_output, bias=self.bias)
+        return output
+
+
+# TODO: add soft prompts
+
+
+class TokensEncoderDecoderModule(MegatronModule):
+    """Token-based (input/output is tokens) encoder-decoder model (e.g. T5 Language model.)"""
+
+    def __init__(
+        self,
+        encoder_arch,
+        decoder_arch,
+        vocab_size,
+        hidden_size,
+        max_position_embeddings,
+        num_layers,
+        num_attention_heads,
+        ffn_hidden_size,
+        apply_query_key_layer_scaling=True,
+        kv_channels=None,
+        num_tokentypes=0,
+        parallel_output=True,
+        pre_process=True,
+        post_process=True,
+        init_method_std=0.02,
+        fp16_cross_entropy=False,
+        use_cpu_initialization=False,
+        hidden_dropout=0.1,
+        precision=16,
+        fp32_residual_connection=False,
+        activations_checkpoint_method=None,
+        activations_checkpoint_num_layers=1,
+        layernorm_epsilon=1e-5,
+        persist_layer_norm=False,
+        bias_gelu_fusion=True,
+        openai_gelu=False,
+        onnx_safe=False,
+        hidden_steps=-1,
+        hidden_blocks=1,
+    ):
+        super(TokensEncoderDecoderModule, self).__init__()
+
+        self.parallel_output = parallel_output
+        self.pre_process = pre_process
+        self.post_process = post_process
+        self.fp16_cross_entropy = fp16_cross_entropy
+        self.precision = precision
+
+        if kv_channels is None:
+            assert (
+                hidden_size % num_attention_heads == 0
+            ), 'hidden_size must be divisible by num_attention_heads if kv_channels is None'
+            kv_channels = hidden_size // num_attention_heads
+
+        self.encoder_embedding = Embedding(
+            hidden_size=hidden_size,
+            vocab_size=vocab_size,
+            max_sequence_length=max_position_embeddings,
+            init_method=init_method_normal(init_method_std),
+            num_tokentypes=num_tokentypes,
+            use_cpu_initialization=use_cpu_initialization,
+            embedding_dropout_prob=hidden_dropout,
+        )
+        self.decoder_embedding = self.encoder_embedding
+        self._encoder_embedding_key = "encoder_embedding"
+        self._decoder_embedding_key = "decoder_embedding"
+
+        encoder = get_encoder_model(
+            arch=encoder_arch,
+            hidden_size=hidden_size,
+            ffn_hidden_size=hidden_size,
```

Comment:
@michalivne This is a bug -- should be `ffn_hidden_size=ffn_hidden_size,` as implemented, it prevents the feedforward layer size from expanding as predicted. 

Here is an example. - for these params:
```
 max_position_embeddings: 512
    hidden_dropout: 0.1
    kv_channels: null
    apply_query_key_layer_scaling: false
    persist_layer_norm: true
    layernorm_epsilon: 1.0e-05
    init_method_std: 0.02
    encoder_arch: transformer
    decoder_arch: transformer
    hidden_size: 36
    ffn_hidden_size: 144
    num_attention_heads: 3
    num_layers: 2
    vocab_size: 128
```

Here are the sizes of the layers:
```
[NeMo I 2022-02-05 20:36:47 test_model:48] enc_dec_model.encoder.model.layers.0.mlp.dense_h_to_4h.weight: torch.Size([36, 36])
[NeMo I 2022-02-05 20:36:47 test_model:48] enc_dec_model.encoder.model.layers.0.mlp.dense_h_to_4h.bias: torch.Size([36])
[NeMo I 2022-02-05 20:36:47 test_model:48] enc_dec_model.encoder.model.layers.0.mlp.dense_4h_to_h.weight: torch.Size([36, 36])
[NeMo I 2022-02-05 20:36:47 test_model:48] enc_dec_model.encoder.model.layers.0.mlp.dense_4h_to_h.bias: torch.Size([36])
```

Expected sizes are:
```
[NeMo I 2022-02-05 20:36:47 test_model:48] enc_dec_model.encoder.model.layers.0.mlp.dense_h_to_4h.weight: torch.Size([144, 36])
[NeMo I 2022-02-05 20:36:47 test_model:48] enc_dec_model.encoder.model.layers.0.mlp.dense_h_to_4h.bias: torch.Size([144])
[NeMo I 2022-02-05 20:36:47 test_model:48] enc_dec_model.encoder.model.layers.0.mlp.dense_4h_to_h.weight: torch.Size([36, 144])
[NeMo I 2022-02-05 20:36:47 test_model:48] enc_dec_model.encoder.model.layers.0.mlp.dense_4h_to_h.bias: torch.Size([36])
```

Reply:
Good catch. FIxed.